### DISCUSSION TITLE
The way of exporting pytorch model including dcn into onnx model.

### DISCUSSION THREAD

**Discussion by YsYusaito** [open-mmlab/mmdetection#9134] [Q&A] [answered]
I want to export pytorch model below into onnx model.
Is it possible to export pytorch model below into onnx model?
If possible, I'd appreciate it if you could tell me the way of exporting.

mmdetection\configs\dcn\mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco.py
[mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth](https://download.openmmlab.com/mmdetection/v2.0/dcn/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth)

**Comment by ZwwWayne** [comment, marked as answer]
The config and models of mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco.py are already supported in [MMDeploy](https://github.com/open-mmlab/mmdeploy)

**Reply by YsYusaito** [reply]
Hi, @ZwwWayne .

Thank you for reply.
I tried exporting model.

・First, I setup mmdetection, mmcv, ...etc.
![image](https://user-images.githubusercontent.com/80863816/199200267-020bcb2d-a853-467a-adcb-6c18403d665b.png)

・Second, I downloaded a model below and put it under the mmdetection directory.(./mmdetection/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth).
[mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth](https://download.openmmlab.com/mmdetection/v2.0/dcn/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth)

・Third, I went to the mmdetection directory and run the following command.
```
python tools/deployment/pytorch2onnx.py \
    "./configs/dcn/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco.py" \
    "mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth" \
    --output-file "mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco.onnx" \
    --input-img "./demo/demo.jpg" \
    --shape 800 1216 \
    --test-img tests/data/color.jpg \
    --verify \
    --simplify
```
However, following error occured.
`RuntimeError: Only tuples, lists and Variables are supported as JIT inputs/outputs. Dictionaries and strings are also accepted, but their usage is not recommended. Here, received an input of unsupported type: numpy.ndarray`

・So, I modified the source codes of mmdetection in order to change inference result from numpy to torch.Tensor.
Modification places are below.
[transforms.py -> bbox2result  l. 131-132]
mmdetection/mmdet/core/transforms.py

※please see # ↑ before modification and # ↑ after modification.
```
def bbox2result(bboxes, labels, num_classes):
    if bboxes.shape[0] == 0:
        return [np.zeros((0, 5), dtype=np.float32) for i in range(num_classes)]
    else:
        if isinstance(bboxes, torch.Tensor):
            # bboxes = bboxes.detach().cpu().numpy()
            # labels = labels.detach().cpu().numpy()
            # ↑ before modification

            bboxes = bboxes
            labels = labels
            # ↑ after modification
        return [bboxes[labels == i, :] for i in range(num_classes)]
```

[fcn_mask_head.py -> get_seg_masks l. 309]
mmdetection/mmdet/models/roi_heads/mask_heads/fcn_mask_head.py

※please see # ↑ before modification and # ↑ after modification.
```
        for inds in chunks:
            masks_chunk, spatial_inds = _do_paste_mask(
                mask_pred[inds],
                bboxes[inds],
                img_h,
                img_w,
                skip_empty=device.type == 'cpu')

            if threshold >= 0:
                masks_chunk = (masks_chunk >= threshold).to(dtype=torch.bool)
            else:
                # for visualization and debugging
                masks_chunk = (masks_chunk * 255).to(dtype=torch.uint8)

            im_mask[(inds, ) + spatial_inds] = masks_chunk

        for i in range(N):
            # cls_segms[labels[i]].append(im_mask[i].detach().cpu().numpy())
            # ↑ before modification
            
            cls_segms[labels[i]].append(im_mask[i])
            # ↑ after modification
        return cls_segms
```

in addition,  I modified following part.
[image_demo.py -> main]
※please see ### modified part ###.

```
def main(args):
    # build the model from a config file and a checkpoint file
    model = init_detector(args.config, args.checkpoint, device=args.device)
    # test a single image
    result = inference_detector(model, args.img)

    ### modified part ###
    result_cpu = []
    count = 0

    for arrs in result:
        result_cpu.append([])
        list_arr = []  # segmentaionの結果を一時的に入れるための配列
        for arr in arrs:
            if(str(type(arr)) == "<class 'torch.Tensor'>"):
                result_cpu[count].append(arr.detach().cpu().numpy())

            elif(str(type(arr)) == "<class 'list'>"):
                arr_list = []

                for tensor_arr in arr:
                    arr_list.append(tensor_arr.detach().cpu().numpy())

                result_cpu[count].append(arr_list)

        count += 1

    result_cpu = tuple(result_cpu)

    show_result_pyplot(
        model,
        args.img,
        result_cpu,
        palette=args.palette,
        score_thr=args.score_thr,
        out_file=args.out_file)
    ######
```

After that, I tried exporting model.
However, same error occured.
`RuntimeError: Only tuples, lists and Variables are supported as JIT inputs/outputs. Dictionaries and strings are also accepted, but their usage is not recommended. Here, received an input of unsupported type: numpy.ndarray`

I used mmdet 2.25.3 and mmcv-full 1.6.2.

Same issue seems to unsolved...
Please help...
https://github.com/open-mmlab/mmdetection/issues/3662

**Reply by YsYusaito** [reply]
Hi, @ZwwWayne 

I tried exporting pytorch model below into onnx model via mmdeploy based on this [page](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/get_started.md).
[mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth](https://download.openmmlab.com/mmdetection/v2.0/dcn/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth)

I used following command.
```
python mmdeploy/tools/deploy.py \
    mmdeploy/configs/mmdet/instance-seg/instance-seg_onnxruntime_static.py \
    mmdetection/configs/dcn/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco.py \
    checkpoints/mask_rcnn_r50_fpn_dconv_c3-c5_1x_coco_20200203-4d9ad43b.pth \
    mmdetection/demo/demo.jpg \
    --work-dir mmdeploy_model/mask-rcnn \
    --device cpu \
    --dump-info
```

However, following error occured.
```
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from mmdeploy_model/mask-rcnn/end2end.onnx failed:Fatal error: MMCVDeformConv2d is not a registered function/op
2022-11-03 07:33:57,344 - mmdeploy - ERROR - visualize onnxruntime model failed.
```

After that, I tried ["Build custom ops"](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/05-supported-backends/onnxruntime.md#build-custom-ops).

However, the problem wasn't solved...
I have already finished [tutorial of model export](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/get_started.md#convert-model) and successfully exported "faster_rcnn_r50_fpn_1x_coco/faster_rcnn_r50_fpn_1x_coco_20200130-047c8118.pth" into onnx.

From the issue below, is it difficult to export pytorch model with MMCVDeformConv2d into onnx?
https://github.com/open-mmlab/mmdetection/discussions/9134#discussioncomment-4033308
